coreblocks/structs_common/csr.py

```
@@ -0,0 +1,278 @@
+from amaranth import *
+
+from coreblocks.transactions import Method, def_method, Transaction
+from coreblocks.utils import assign
+from coreblocks.params.genparams import GenParams
+from coreblocks.params.layouts import FuncUnitLayouts, CSRLayouts
+from coreblocks.params.isa import Funct3
+
+
+class CSRRegister(Elaboratable):
```

The current design of `CSRRegister` basically decouples CSR reads/writes from the modules handling them, and I don't know if that is a good idea. In particular, I believe it's now impossible for CSR side effects to have latency. The `_fu_read` and `_fu_write` methods are always ready, and there is no way to make the CSR unit wait if the operation cannot be performed in a single clock cycle. Also, in the example code in the docstring below, if there is some latency in the transaction, information about side effects will be silently lost. In other words, this design is not latency-insensitive.